### PR TITLE
Move update tests to compat tests

### LINKED ISSUE #1610: Move update tests to compat
### What should be done?

Move update tests to compat, this will help us not having update tests assertion for every new collection added to shareddata.

Skip tests (and create new issues for them) that can't be fixed quickly.

### Where?

* https://github.com/FerretDB/FerretDB/blob/main/integration/update_test.go
* https://github.com/FerretDB/FerretDB/blob/main/integration/update_field_test.go

### PR DESCRIPTION
**Pull request by chilagrow**
# Description

Closes #1610.

## Readiness checklist

* [ ] I added tests for new functionality or bugfixes.
* [ ] I ran `task all`, and it passed.
* [x] I added/updated comments for both exported and unexported top-level declarations (functions, types, etc).
* [ ] I checked comments rendering with `task godocs`.
* [x] I ensured that the title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

File: integration/update_field_test.go [hunk 1]
```diff
@@ -279,299 +278,6 @@ func TestUpdateFieldCurrentDate(t *testing.T) {
...
-				id:       "double",
-				update:   bson.D{{"$inc", bson.D{{"v", int32(1)}}}},
-				expected: bson.D{{"_id", "double"}, {"v", float64(43.13)}},
-			},
-			"DoubleLongIncrement": {
-				id:       "double",
-				update:   bson.D{{"$inc", bson.D{{"v", int64(1)}}}},
-				expected: bson.D{{"_id", "double"}, {"v", float64(43.13)}},
-			},
-			"DoubleDoubleMaxIncrement": {
```
Comment (chilagrow) [inline integration/update_field_test.go:326]: This test was removed, it produced `+INF` and we no longer support it.

File: integration/helpers.go [hunk 2]
```diff
@@ -184,6 +184,27 @@ func AssertMatchesCommandError(t *testing.T, expected, actual error) {
 	assert.Equal(t, eErr.Code, aErr.Code)
 }
 
+// AssertMatchesWriteErrorCode asserts error code.
+func AssertMatchesWriteErrorCode(t *testing.T, expected, actual error) {
+	t.Helper()
+
+	//nolint: errorlint // using errors.As does not work for mongo.WriteException
+	aErr, ok := actual.(mongo.WriteException)
```
Comment (chilagrow) [inline integration/helpers.go:192]: Ignoring linter, this didn't work.

```
	var aErr, eErr mongo.WriteException

	if ok := errors.As(actual, aErr); !ok || len(aErr.WriteErrors) != 1 {
		assert.Equal(t, expected, actual)
		return
	}
```

nor this
```
	var aErr, eErr *mongo.WriteException

	if ok := errors.As(actual, &aErr); !ok || len(aErr.WriteErrors) != 1 {
		assert.Equal(t, expected, actual)
		return
	}
```

File: integration/update_compat_test.go [hunk 3]
```diff
@@ -143,6 +151,254 @@ func testUpdateCompat(t *testing.T, testCases map[string]updateCompatTestCase) {
...
+	skip       string                   // skips test if non-empty
+	paths      []types.Path             // paths to check after update
+	update     bson.D                   // required
+	filter     bson.D                   // defaults to bson.D{{"_id", id}}
+	resultType compatTestCaseResultType // defaults to nonEmptyResult
+}
+
+// testUpdateCompat tests update compatibility test cases for current date.
+// It checks current date in compat and target are within acceptable difference.
+func testUpdateCurrentDateCompat(t *testing.T, testCases map[string]updateCurrentDateCompatTestCase) {
```
Comment (w84thesun) [inline integration/update_compat_test.go:306]: I can't find any `currentDate` test cases. Is this for the future?

**Review by w84thesun — COMMENTED** [review]
Asked about `testUpdateCurrentDateCompat`. I noticed Tigris tests are failing. Maybe some `skip` should be added?

File: internal/types/path.go [hunk 4]
```diff
@@ -220,6 +220,13 @@ func insertByPath(doc *Document, path Path) error {
 						FormatAnyValue(v),
 					)
 				}
+			default:
+				return fmt.Errorf(
```
Comment (rumyantseva) [inline internal/types/path.go:224]: Should this function be called for anything but doc/array at all?

File: internal/handlers/common/update.go [hunk 5]
```diff
@@ -143,6 +143,17 @@ func processSetFieldExpression(doc, setDoc *types.Document, setOnInsert bool) (b
 	for _, setKey := range setDocKeys {
 		setValue := must.NotFail(setDoc.Get(setKey))
 
+		if setOnInsert {
+			// $setOnInsert do not set null and empty array value.
```
Comment (rumyantseva) [inline internal/handlers/common/update.go:147]: 👍

**Review by rumyantseva — COMMENTED** [review]
I started reviewing things, I'll need more time.

File: integration/update_compat_test.go (hunk 3 above)
Comment (chilagrow) [reply integration/update_compat_test.go:306]: Please see the change in [integration/update_field_compat_test.go](https://github.com/FerretDB/FerretDB/pull/1659/files#diff-02beb83969c6feb8b48d15aebcd299647e0aa2e8b1e5c531e74eb2c4f636ad85)

There are a lot of changes in the file and Github does `Large diffs are not rendered by default.` :see_no_evil:

File: internal/types/path.go (hunk 4 above)
Comment (chilagrow) [reply internal/types/path.go:224]: Please suggest where is a better to add this check :hugs:

I don't know the answer to _if this function should be called for non-doc/array_, but it _is_ being called for non-doc/array and we need to return error.

Comment (rumyantseva) [reply internal/types/path.go:224]: Ok, I think I got the idea (sort of). I'll ask in person in a call not to extend the current thread.

**Comment by chilagrow** [comment]
Tests are taking long. I'm surprised it didn't timeout given postgres took 12m.

image

**Comment by chilagrow** [comment]
I marked PR `do-not-merge` because I'm hitting 10min timeout on test every so frequently. Just happened on postgres too.

File: integration/update_compat_test.go (hunk 3 above)
Comment (w84thesun) [reply integration/update_compat_test.go:306]: Oh, right, thanks!

**Comment by AlekSi** [comment]
> I'm surprised it didn't timeout given postgres took 12m.

That's the whole job run time; it includes building images, starting environment, etc.

**Review by w84thesun — DISMISSED** [review]

File: integration/helpers.go (hunk 2 above)
Comment (noisersup) [reply integration/helpers.go:192]: Both examples provide to `errors.As` a value or a pointer to the pointer.
Please consider something like this (it passed test-integration-pg so I assume it's working):
```go
var aErr, eErr mongo.WriteException

if ok := errors.As(actual, &aErr); !ok || len(aErr.WriteErrors) != 1 {
	assert.Equal(t, expected, actual)
	return
}

if ok := errors.As(expected, &eErr); !ok || len(eErr.WriteErrors) != 1 {
	assert.Equal(t, expected, actual)
	return
}

```

File: integration/helpers.go [hunk 6]
```diff
@@ -184,6 +184,27 @@ func AssertMatchesCommandError(t *testing.T, expected, actual error) {
 	assert.Equal(t, eErr.Code, aErr.Code)
 }
 
+// AssertMatchesWriteErrorCode asserts error code.
```
Comment (noisersup) [inline integration/helpers.go:187]:
```suggestion
// AssertMatchesWriteErrorCode asserts error codes are the same.
```

**Review by noisersup — CHANGES_REQUESTED** [review]
I left few comments. Apart from them everything looks good!

**Review by rumyantseva — DISMISSED** [review]
The changes and introduced tests look reasonable to me.

File: integration/helpers.go (hunk 6 above)
Comment (chilagrow) [reply integration/helpers.go:187]: thanks!

File: integration/helpers.go (hunk 2 above)
Comment (chilagrow) [reply integration/helpers.go:192]: yea just saw that :see_no_evil: thanks.

**Review by rumyantseva — APPROVED** [review]
LGTM

**Review by noisersup — APPROVED** [review]

**Pull request merged**